### PR TITLE
Webdriver factory implements webdriver provider

### PR DESCRIPTION
**Pull request by asolntsev**
## Proposed changes
Allow users to extend `ChromeDriverFactory` etc. instead of `WebDriverProvider`. 
So they get all the built-in Selenide settings and proxy, and add their specific chrome options/arguments.

It fixes issues #1097 and #676.

## Checklist
- [x] Checkstyle and unit tests are passed locally with my changes by running `gradlew check chrome_headless firefox_headless` command
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: src/main/java/com/codeborne/selenide/webdriver/ChromeDriverFactory.java [hunk 1]
```diff
@@ -2,88 +2,115 @@
...
-  ChromeOptions createChromeOptions(Config config, Proxy proxy) {
+  @Override
+  public void setupBinary() {
+    if (isSystemPropertyNotSet("webdriver.chrome.driver")) {
+      WebDriverManager.chromedriver().setup();
+    }
+  }
+
+  @Override
+  @SuppressWarnings("deprecation")
```
Comment (dstekanov) [inline src/main/java/com/codeborne/selenide/webdriver/ChromeDriverFactory.java:48]: Will it be good for Selenium 4 migration?

File: src/main/java/com/codeborne/selenide/webdriver/MergableCapabilities.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A subclass of MutableCapabilities which has fixed `merge` method:
+ * it can properly merge all these ChromeOptions etc. with their Maps inside of Maps.
+ */
+public class MergableCapabilities extends MutableCapabilities {
```
[the remark this answers is not in the document]
Comment (asolntsev) [reply src/main/java/com/codeborne/selenide/webdriver/MergableCapabilities.java:15]: fixed

File: src/main/java/com/codeborne/selenide/webdriver/ChromeDriverFactory.java (hunk 1 above)
Comment (asolntsev) [reply src/main/java/com/codeborne/selenide/webdriver/ChromeDriverFactory.java:48]: When Selenium 4 gets finally released, we will find a solution. :) 

Actually I am going to suggest a fix to `ChromeOptions` class in Selenium 4.

File: src/test/java/com/codeborne/selenide/webdriver/MergableCapabilitiesTest.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+import static java.util.Arrays.asList;
+import static java.util.Collections.emptyList;
+import static org.assertj.core.api.Assertions.assertThat;
+
+class MergableCapabilitiesTest {
+  @Test
+  void mergesChromeOptions() {
+    ChromeOptions base = new ChromeOptions();
+    base.addArguments("--a", "--b");
```
[the remark this answers is not in the document]
Comment (asolntsev) [reply src/test/java/com/codeborne/selenide/webdriver/MergableCapabilitiesTest.java:15]: Good point!
I have added merging of arrays too.

I have also added such a test:

```java
    ChromeOptions base = new ChromeOptions();
    base.setExperimentalOption("excludeSwitches", new String[]{"foo"});
    ChromeOptions extra = new ChromeOptions();
    extra.setExperimentalOption("excludeSwitches", new String[]{"bar", "zzz"});

    MergeableCapabilities result = new MergeableCapabilities(base, extra);
     assertThat(result....get("excludeSwitches")).isEqualTo(new String[]{"foo", "bar", "zzz"});
```

File: src/main/java/com/codeborne/selenide/webdriver/ChromeDriverFactory.java [hunk 4]
```diff
@@ -2,88 +2,115 @@
...
+  private static final Pattern REGEX_REMOVE_QUOTES = Pattern.compile("\"", Pattern.LITERAL);
 
   @Override
   boolean supports(Config config, Browser browser) {
     return browser.isChrome();
   }
 
-  ChromeOptions createChromeOptions(Config config, Proxy proxy) {
+  @Override
+  public void setupBinary() {
```
[the remark this answers is not in the document]
Comment (asolntsev) [reply src/main/java/com/codeborne/selenide/webdriver/ChromeDriverFactory.java]: Good point! Renamed.

**Pull request merged**